iree/hal/command_buffer.c

```diff
@@ -32,13 +32,31 @@ IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_command_buffer_create(
   IREE_ASSERT_ARGUMENT(out_command_buffer);
   *out_command_buffer = NULL;
   IREE_TRACE_ZONE_BEGIN(z0);
+
+  if (iree_all_bits_set(mode,
+                        IREE_HAL_COMMAND_BUFFER_MODE_ALLOW_INLINE_EXECUTION)) {
+    // Inline command buffers must be one-shot and primary.
+    if (!iree_all_bits_set(mode, IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT)) {
+      return iree_make_status(
+          IREE_STATUS_INVALID_ARGUMENT,
+          "inline command buffers must be one-shot and primary");
+    }
+  }
```

Comment:
I debated with that, but as you noted below this is more of an API contract on object creation rather than what is in _validation.c - that's focused entirely on validating command buffer recording calls and where 99% of the potential overhead comes from - we should (ideally) be creating a command buffer once at startup and reusing it forever, and one bitwise comparison is not worth bothering about :)